WIP: Database and Network interaction.

Kostyshina opened the pull request:
Added MVP architecture to display items from db (without recyclerView and images in it, and without button animation)

Kostyshina committed:
Added BaseActivity with navigationDrawer and Toolbar.

Kostyshina committed:
MVP architecture was added (simple example, Presenter creating model element by itself).

Kostyshina committed:
Added db for the application. No saving Presenter is provided.

giffell commented on app/src/main/AndroidManifest.xml:
```diff
@@ -11,7 +11,10 @@
-        android:theme="@style/AppTheme">
+        android:theme="@style/AppTheme.NoActionBar">
+
+        <activity android:name=".activity.BaseActivity"/>
```
А зачем это тут?

giffell commented on app/src/main/java/com/andersenlab/boilerplate/BoilerplateApp.java:
```diff
@@ -43,5 +45,14 @@ public void onCreate() {
+        dbHelper.deleteAllImages();
+        String[] images = new String[]{"Image 1", "Image 2", "Image 3", "Image 4",
+                "Image 5", "Image 6", "Image 7", "Image 8", "Image 9", "Image 10",
+                "Image 11", "Image 12", "Image 13", "Image 14", "Image 15", "Image 16"};
```
В ресурсы бы вынести

giffell commented on app/src/main/java/com/andersenlab/boilerplate/activity/BaseActivity.java:
```diff
@@ -0,0 +1,111 @@
+/**
+ * Base class for all activities.
+ */
+
```
Итак понятно что это базовый класс Активити из названия

rmazheika commented on app/src/main/java/com/andersenlab/boilerplate/activity/MainActivity.java:
```diff
@@ -2,26 +2,61 @@
+
+    private static final String CURRENT_ITEM_BUNDLE_KEY = "com.andersenlab.boilerplate.activity.currentItem";
+
+    @BindView(R.id.tv_main_hello_world) TextView tvHelloWorld;
```
Бинд вьюх делай ниже приватных переменных, констант и конструкторов

Kostyshina replied:
@rmazheika а почему ниже конструкторов ?
есть например такой конвеншн https://web.archive.org/web/20130516014426/http://www.oracle.com/technetwork/java/codeconventions-141855.html#1852 - тут методы по логике, а не по видимости расставляются. что так же неплохая практика
в настройках студии точно так же, иннер классы в самом низу

rmazheika replied:
А какой смысл биндить вьюхи до конструктора, который этот класс по сути создает? Непоследовательно

Kostyshina replied:
@rmazheika окей, логика мне понятна. но при этом нарушается конвеншн, что сначала поля, потом методы. я еще поищу, может найду что по этому поводу

giffell commented on app/src/main/java/com/andersenlab/boilerplate/activity/BaseActivity.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    protected boolean useToolbar()
+    {
```
что за синтаксис языка `C/C++` ?

Kostyshina replied:
@giffell ты про то, что curly braces на следующей строчке ?

giffell replied:
угу

rmazheika commented on app/src/main/java/com/andersenlab/boilerplate/activity/MainActivity.java:
```diff
@@ -2,26 +2,61 @@
         ButterKnife.bind(this);
+
+        if (savedInstanceState != null) {
+            currentItem = savedInstanceState.getParcelable(CURRENT_ITEM_BUNDLE_KEY);
```
А если Parcelable null? Будет NPE

Kostyshina replied:
@rmazheika что выкенет NPE ? null вернет и все. я везде на null проверяю, где юзаю currentItem.

rmazheika replied:
Хорошо

giffell replied:
Прелесть в том, чтобы наоборот избавиться от проверок на `null`

giffell commented on app/src/main/java/com/andersenlab/boilerplate/activity/BaseActivity.java:
```diff
@@ -0,0 +1,111 @@
+        super.setContentView(R.layout.activity_base);
+
+        drawerLayout = findViewById(R.id.base_root_layout);
+        toolbar = findViewById(R.id.toolbar);
```
Ты же говорила `ButterKnife` подключала

Kostyshina replied:
@giffell я то подключала, но так как в MainActivity вызывается onCreate() BaseActivity, то у меня там ругался ButterKnife.bind(this) ругался, что не мог найти в лэйауте бэйза вьюху из activity_main. но я посмотрю, первое что мне пришло в голову просто

Kostyshina replied:
@giffell я поняла как сделать через bind view. в наследниках bind'а не будет, а в BaseActivity так:

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }
    @Override
    public void setContentView(@LayoutRes int layoutResID) {
        super.setContentView(R.layout.activity_base);
        FrameLayout activityContainer = getContainerLayout();
        getLayoutInflater().inflate(layoutResID, activityContainer, true);
        ButterKnife.bind(this);
        initNavDrawer();
    }
    private FrameLayout getContainerLayout() {
        return (FrameLayout) findViewById(R.id.activity_container);
    }

но тогда все равно от findViewById не избавиться

giffell replied:
Почему? возьми верни вью которая будет байндиться

rmazheika commented on app/src/main/java/com/andersenlab/boilerplate/activity/MainActivity.java:
```diff
@@ -2,26 +2,61 @@
 
     @OnClick(R.id.btn_main_add_item)
     public void addItem(View view) {
         Timber.i("Tap on add item button");
```
Не ошибка, но логи я бы в конце писал, если кода мало.
Так воспринимается лучше. Дублировать по аналогичным случаям не буду

rmazheika commented on app/src/main/java/com/andersenlab/boilerplate/activity/MainActivity.java:
```diff
@@ -33,4 +68,26 @@ protected boolean useDrawerToggle() {
+        if (mainPresenter != null) {
+            int itemNum = 1;
+            if (currentItem != null)
+                itemNum += currentItem.getSerialNumber() + increment;
```
Как-то криво смотрится, поменяй на Number, потому что непонятно, что за инкремент для загрузки

rmazheika commented on app/src/main/java/com/andersenlab/boilerplate/presenter/MainPresenter.java:
```diff
@@ -0,0 +1,21 @@
+            else
+                getMvpView().showNewItem(
+                        new Image(serialNum,
+                                BoilerplateApp.getInstance()
```
Печальный код, я бы передавал контекст по крайней мере

rmazheika commented on app/src/main/java/com/andersenlab/boilerplate/view/MainMvpView.java:
```diff
@@ -0,0 +1,8 @@
+
+public interface MainMvpView extends MvpView {
+    void showNewItem(Image item);
+    void showError();
```
Пример - это хорошо, но как различать ошибки? Хотя бы сообщение в параметры передавай, а то не универсально получается.

rmazheika commented on app/src/main/java/com/andersenlab/boilerplate/activity/BaseActivity.java:
```diff
@@ -0,0 +1,111 @@
+    }
+
+    public String getToolbarTitle(){
+        return "";
```
А че это такое?

Kostyshina replied:
@rmazheika ну это на будущее, я сделала NavigationDrawer в BaseActvity и toolbar. наследники буду переопределять в зависимости о того, какой у них тайтл у тулбара и юзать тогл или нет.

rmazheika replied:
Коммент добавь в таком случае

Kostyshina committed:
Added recyclerView with textView items and created needed ImageAdapter. Review changes: removed BaseActivity from manifest; placed string array to string resources; added custom exception for MvpView; ButterKnife.bind placed to BaseActivity; added some JavaDoc.

Kostyshina committed:
Replaced textView item with imageView in recyclerView. Load images with glide implemented. Design of button addItem was applied.

Kostyshina committed:
Created custom bottom view for button in MainActivity. Also created CoordinatorLayout.Behavior to occur expanding/collapsing animation on recyclerView scrolling.

Kostyshina commented:
Added recyclerView with Image items, which are loaded with Glide, also added scrolling aware in bottom bar (layout_behavior for CoordinatorLayout) and button ripple style.

giffell commented on app/src/main/res/layout/activity_base.xml:
```diff
@@ -0,0 +1,44 @@
+                style="@style/TextAppearance.AppCompat.Widget.ActionBar.Title"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                tools:text="@string/app_name" />
```
А избавиться от этого поля нельзя разве?

Kostyshina replied:
@giffell от textView ? ну можно, это я на всякий. на всех проектах надпись у тулабара всегда по середине была, да и размер шрифта тоже был больше стандартного. могу убрать и сетать стандартный тайтл тулбара.

giffell replied:
На всякий случай не надо

giffell commented on app/src/main/java/com/andersenlab/boilerplate/activity/MainActivity.java:
```diff
@@ -1,27 +1,119 @@
+    private static final String IMAGES_BUNDLE_KEY = "com.andersenlab.boilerplate.activity.imageList";
+
+    @BindView(R.id.rv_main_images) RecyclerView rvImages;
+    @BindView(R.id.fbs_main_add_item) FloatingBottomSheet fbsAddItem;
```
ну и названия, ужс просто

giffell commented on app/src/main/java/com/andersenlab/boilerplate/activity/MainActivity.java:
```diff
@@ -1,27 +1,119 @@
     @Override
     protected void onCreate(Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
+        Timber.i("onCreate");
```
на сколько нужны эти сообщения?

giffell commented on app/src/main/java/com/andersenlab/boilerplate/activity/MainActivity.java:
```diff
@@ -1,27 +1,119 @@
+        mainPresenter.detachView();
+    }
+
+    @OnClick(R.id.fbs_main_add_item)
```
наверное еще хуже это через xml прокидывать

giffell commented on app/src/main/java/com/andersenlab/boilerplate/activity/MainActivity.java:
```diff
@@ -1,27 +1,119 @@
+
+    @OnClick(R.id.fbs_main_add_item)
     public void addItem(View view) {
+        loadItem();
```
не вижу тут addItem никакого

giffell commented on app/src/main/java/com/andersenlab/boilerplate/activity/MainActivity.java:
```diff
@@ -1,27 +1,119 @@
+
+    @Override
+    public void showNewItem(Image item) {
+        setNewItem(item);
```
так Set- или Show- newItem ?

giffell commented on app/src/main/java/com/andersenlab/boilerplate/activity/MainActivity.java:
```diff
@@ -1,27 +1,119 @@
+    }
+
+    @Override
+    public void showEmpty() {
```
не информативное название

giffell commented on app/src/main/java/com/andersenlab/boilerplate/activity/MainActivity.java:
```diff
@@ -1,27 +1,119 @@
+    }
+
+    private void loadItem() {
+        if (mainPresenter != null) {
```
а что будет если `null`?

giffell commented on app/src/main/java/com/andersenlab/boilerplate/activity/MainActivity.java:
```diff
@@ -1,27 +1,119 @@
+    }
+
+    private void setNewItem(Image item) {
+        if (imageList == null)
```
у тебя он же в самом начале создается, а потом затираешь на null? Неоднозначное поведение

giffell commented on app/src/main/java/com/andersenlab/boilerplate/activity/MainActivity.java:
```diff
@@ -1,27 +1,119 @@
+
+        imageList.add(item);
+        imageAdapter.addItem(item);
+        rvImages.scrollToPosition(imageList.size() - 1);
```
а если будет пустой?

Kostyshina replied:
@giffell я же перед этим один элемент добавляю

giffell replied:
тогда зачем скролить если у тебя 1 элемент?

giffell commented on app/src/main/java/com/andersenlab/boilerplate/adapter/ImageAdapter.java:
```diff
@@ -0,0 +1,95 @@
+
+    @Override
+    public void onViewRecycled(ImageViewHolder holder) {
+        holder.clearContent();
```
Glide это делает автоматически

Kostyshina replied:
@giffell Не делает, я прочитала это у них в вики https://bumptech.github.io/glide/doc/getting-started.html

giffell commented on app/src/main/java/com/andersenlab/boilerplate/adapter/viewholder/BaseViewHolder.java:
```diff
@@ -0,0 +1,23 @@
+     * Abstract method for binding views by modelItems.
+     * @param modelItem Instance of class, that extends {@link BaseModel} abstract class.
+     */
+    public abstract void bind(final T modelItem);
```
полезность метода непонятна, мы просто дергаем его из `onBindViewHolder` не более

Kostyshina replied:
@giffell а что еще он должен делать ? это же базовый абстрактный класс, будет например у меня адептер с несколькими вьюхолдерами. я же не буду все сеты писать прямо в onBindView в зависимости от вьюхолдера например

giffell replied:
ты передаешь модель в метод и ей данные сетаешь не более. Поэтому вынос логики в метод не вижу

giffell commented on app/src/main/java/com/andersenlab/boilerplate/customview/FloatingBottomSheet.java:
```diff
@@ -0,0 +1,110 @@
+        ButterKnife.bind(this, view);
+
+        flButtonContainer.setBackgroundColor(a.getColor(R.styleable.FloatingBottomSheet_android_background,
+                ContextCompat.getColor(context, android.R.color.white)));
```
через ресурсы нельзя ли это сделать?

Kostyshina replied:
@giffell да, окей, не подумала

giffell commented on app/src/main/java/com/andersenlab/boilerplate/customview/FloatingBottomSheet.java:
```diff
@@ -0,0 +1,110 @@
+            constrParams.gravity = Gravity.BOTTOM;
+            constrParams.setBehavior(behavior);
+            this.setLayoutParams(constrParams);
+        }
```
костыли какие-то, через кастовая вьюшка для кнопки и внутри нее еще behavior сетается.
Как вижу я это я. Какой-нибудь ViewGroup с фоном и внутри кнопка по центру. Написать behavior и заседать его через xml

giffell replied:
Мне кажется ты немного перестаралась :)

Kostyshina replied:
@giffell просто мне хотелось, чтобы у кастомной вьюхи был этот behavior по дефолту, но так как мне нужны атрибуты в самом behavior - то я не могла сделать это через аннотацию (вызывается дефолтный конструктор). тип как свой вариант FloatingButton. оно конечно сейчас так сделано, что другой behavior не засетается, да и состояние его я пока не придумала как сохранять в таком виде. я еще буду думать как лучше сделать.

giffell commented on app/src/main/java/com/andersenlab/boilerplate/model/db/DatabaseHelper.java:
```diff
@@ -0,0 +1,151 @@
+ * Helper class to work with database.
+ */
+
+public class DatabaseHelper extends SQLiteOpenHelper {
```
А теперь добавь Realm. И замени его использование в проекте. Sqlite пусть будет, мало ли пригодится :)

giffell replied:
Потом скажешь на сколько удобно все сделано для замены

Kostyshina replied:
@giffell ну раз ты упомянул - то наверное не удобно )

giffell replied:
Хотим синхронизацию между двумя базами данных @rmazheika

Kostyshina committed:
Review changes: java naming conversation; no OnClick annotations; method naming. Set DefaultBehavior for custom view (no saving state after recreation). Added signing for release flavor.

Kostyshina committed:
Fixed saving view collapsed/expanded state on view recreation.

Kostyshina committed:
Added appbar_scroll behavior for toolbar hiding/showing on scrolling. Set bottom bar dependency on AppBarLayout. Removed textView from toolbar.

Kostyshina committed:
Added CollapsingToolbarLayout to activity_main.

Kostyshina committed:
Integrated Realm to the application.

Kostyshina committed:
Added unit tests for SQLite database and realm.

Kostyshina committed:
Review changes: tests were renamed; javadoc was corrected; replaced if's with assert's.

giffell commented on app/src/androidTest/java/com/andersenlab/boilerplate/RealmImageInteractorInstrumentalTest.java:
```diff
@@ -0,0 +1,106 @@
+
+        String imageUrl = getImageUrl();
+
+        assertNotNull("All string urls from resources are null", imageUrl);
```
бесполезный тест, ты сама руками задаешь значения в ресурсах

Kostyshina replied:
@giffell в данном случае да, но когда подкрутится получения картинок с сервера - там уже не смогу гарантировать.

giffell commented on app/src/androidTest/java/com/andersenlab/boilerplate/RealmImageInteractorInstrumentalTest.java:
```diff
@@ -0,0 +1,106 @@
+    public void deleteImagesFromRealm() {
+        realmInteractor.clearAll(Image.class);
+
+        assertFalse("Cleaning realm failed", realmInteractor.hasObjects(Image.class));
```
чтобы что-то удалить, нужно туда что-то положить

giffell commented on app/src/androidTest/java/com/andersenlab/boilerplate/RealmImageInteractorInstrumentalTest.java:
```diff
@@ -0,0 +1,106 @@
+
+        Image imageFromRealm = realmInteractor.getObjectById(Image.class, 1);
+        assertTrue("Adding to realm failed", imageFromRealm != null &&
+                imageUrl.equals(imageFromRealm.getImageUrl()));
```
сравнение всего объекта надо делать, а не только конкретного поля. И тем тем более есть ассерт для сравнения значений

Kostyshina replied:
@giffell я бы с радостью сделала стандартный equals, но то что возвращает realm - это ссылка на объект в нем, а не в хипе. так что буду делать по полям

giffell commented on app/src/androidTest/java/com/andersenlab/boilerplate/RealmImageInteractorInstrumentalTest.java:
```diff
@@ -0,0 +1,106 @@
+
+        int i = 0;
+        for (String imageUrl : images) {
+            i++;
```
что это?

Kostyshina committed:
Started integrating Retrofit and RxJava to the project.

Kostyshina committed:
UI test was added for recyclerView in MainActivity. Review changes: adding element to db before checking deleting operation, some refactor.

Kostyshina committed:
Updated gradle plugin version to 3.0.1. Some libraries were updated.

Kostyshina committed:
Some Studio warnings were corrected. Right implementation of GlideModule.

Kostyshina committed:
Glide Preloading API for recyclerView have been implemented. Trying loading images through okhttp3 (not work well). Changed context in ImageAdapter constructor parameters set to RequestManager.

Kostyshina committed:
Moved recyclerView to separate fragment.

Kostyshina committed:
Added navigationMenu and backStack. Starting fixing standard AppBarLayout.Behavior (it's scrolling when not items in recyclerView is added)

Kostyshina committed:
AppBarLayout.Behavior standard canDragView was disabled.

Kostyshina commented:
Upgraded some libs. Added Fragments. Changed AppBarLayout behavior.

Kostyshina committed:
Implemented uploading images through Retrofit on separate tab. Code refactoring.

Kostyshina committed:
Logic of loading images from network was changed. Added disabling addItem button. Saving state on activity recreation in process of network loading.

Kostyshina committed:
Fixed some critical bugs: addItem button had the same state for all fragments; incorrect communication with Activity from fragment. Added CachedObservable for caching request and retrieving response.

Kostyshina commented:
Added loading images through Retrofit on separate tab.
